drivers/can/mcp2515.c

```diff
@@ -0,0 +1,748 @@
+			    u8_t buf_len)
+{
+	u8_t opcode_buf[2];
+
+	opcode_buf[0] = MCP2515_OPCODE_READ;
+	opcode_buf[1] = reg_addr;
+
+	struct spi_buf tx_buf[2];
+
+	tx_buf[0].buf = opcode_buf;
+	tx_buf[0].len = 2;
+
+	tx_buf[1].buf = NULL;
+	tx_buf[1].len = buf_len;
+
+	const struct spi_buf_set tx = {
+		.buffers = tx_buf,
+		.count = 2
+	};
+
+	struct spi_buf rx_buf[2];
+
+	rx_buf[0].buf = NULL;
+	rx_buf[0].len = 2;
+
+	rx_buf[1].buf = buf_data;
+	rx_buf[1].len = buf_len;
+
+	const struct spi_buf_set rx = {
+		.buffers = rx_buf,
+		.count = 2
+	};
+
+	return spi_transceive(DEV_DATA(dev)->spi, &DEV_DATA(dev)->spi_cfg,
+			      &tx, &rx);
+}
+
+const int mcp2515_request_opmode(struct device *dev, u8_t mode)
+{
+	u8_t canstat;
+
+	mcp2515_bit_modify(dev, MCP2515_ADDR_CANCTRL, 0x07 << 5, mode << 5);
+	mcp2515_read_reg(dev, MCP2515_ADDR_CANSTAT, &canstat, 1);
+
+	if ((canstat >> 5) != mode) {
+		SYS_LOG_ERR("Failed to set MCP2515 operation mode");
+		return -EIO;
+	}
+
+	return 0;
+}
+
+static u8_t mcp2515_get_mcp2515_opmode(enum can_mode mode)
+{
+	switch (mode) {
+	case CAN_NORMAL_MODE:
+		return MCP2515_MODE_NORMAL;
+	case CAN_SILENT_MODE:
+		return MCP2515_MODE_SILENT;
+	case CAN_LOOPBACK_MODE:
+		return MCP2515_MODE_LOOPBACK;
+	default:
+		SYS_LOG_ERR("Unsupported CAN Mode %u", mode);
+		return MCP2515_MODE_SILENT;
+	}
+}
+
+static void mcp2515_convert_can_msg_to_mcp2515_frame(
+	const struct can_msg *source, u8_t *target)
+{
+	if (source->id_type == CAN_STANDARD_IDENTIFIER) {
+		target[MCP2515_FRAME_OFFSET_SIDH] = source->std_id >> 3;
+		target[MCP2515_FRAME_OFFSET_SIDL] = (source->std_id & 0x07) << 5;
+	} else {
+		target[MCP2515_FRAME_OFFSET_SIDH] = source->ext_id >> 21;
+		target[MCP2515_FRAME_OFFSET_SIDL] =
+			(((source->ext_id >> 18) & 0x07) << 5) | (BIT(3)) |
+			((source->ext_id >> 16) & 0x03);
+		target[MCP2515_FRAME_OFFSET_EID8] = source->ext_id >> 8;
+		target[MCP2515_FRAME_OFFSET_EID0] = source->ext_id;
+	}
+
+	u8_t rtr = 0;
+
+	if (source->rtr == CAN_REMOTEREQUEST) {
+		rtr = BIT(6);
+	}
+
+	u8_t dlc = (source->dlc) & 0x0F;
+
+	target[MCP2515_FRAME_OFFSET_DLC] = rtr | dlc;
+
+	u8_t data_idx = 0;
+
+	for (; data_idx < 8; data_idx++) {
+		target[MCP2515_FRAME_OFFSET_D0 + data_idx] = source->data[data_idx];
+	}
+}
+
+static void mcp2515_convert_mcp2515_frame_to_can_msg(const u8_t *source,
+						     struct can_msg *target)
+{
+	if (source[MCP2515_FRAME_OFFSET_SIDL] & BIT(3)) {
+		target->id_type = CAN_EXTENDED_IDENTIFIER;
+		target->ext_id = (source[MCP2515_FRAME_OFFSET_SIDH] << 21) |
+				 ((source[MCP2515_FRAME_OFFSET_SIDL] >> 5) << 18) |
+				 ((source[MCP2515_FRAME_OFFSET_SIDL] & 0x03) << 16) |
+				 (source[MCP2515_FRAME_OFFSET_EID8] << 8) |
+				 source[MCP2515_FRAME_OFFSET_EID0];
+
+	} else {
+		target->id_type = CAN_STANDARD_IDENTIFIER;
+		target->std_id = (source[MCP2515_FRAME_OFFSET_SIDH] << 3) |
+				 (source[MCP2515_FRAME_OFFSET_SIDL] >> 5);
+	}
+	target->dlc = source[MCP2515_FRAME_OFFSET_DLC] & 0x0F;
```

Comment:
add an empty line after '}' unless it's another } etc... Apply that everywhere.

Reply:
Done